### PR TITLE
Change updateServiceProvider method for SAML configs

### PR DESCRIPTION
**Pull request by hwupathum**
### Proposed changes in this pull request
Change updateServiceProvider method  for SAML configs by adding currentIssuer parameter to check if the issuer exists in update operation

### Related PR
- https://github.com/wso2/carbon-identity-framework/pull/4753

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java [hunk 1]
```diff
@@ -517,7 +522,12 @@ public boolean updateServiceProvider(SAMLSSOServiceProviderDO serviceProviderDO)
             if (!isTransactionStarted) {
                 registry.beginTransaction();
             }
-            registry.put(path, resource);
+            // Delete the current resource if the issuer value is updated.
+            if (isIssuerUpdated) {
```
Comment (madurangasiriwardena) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java:526]: Shouldn't we create the new resource before deleting the old one? Can you validate the behaviour with the old implementation?

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java [hunk 2]
```diff
@@ -517,7 +522,12 @@ public boolean updateServiceProvider(SAMLSSOServiceProviderDO serviceProviderDO)
...
+                registry.delete(currentPath);
+            }
+            // Update the current resource. If the resource does not exist, it will be created.
```
Comment (madurangasiriwardena) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java:529]:
```suggestion
            // Update the resource.
            // If the issuer is updated, new resource will be created.
            // If the issuer is not updated, existing resource will be updated.
```

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java (hunk 1 above)
Comment (hwupathum) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java:526]: In the old implementation. Delete method is called before the add method.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/SAMLSSOServiceProviderManager.java [hunk 3]
```diff
@@ -72,16 +72,17 @@ public boolean addServiceProvider(SAMLSSOServiceProviderDO serviceProviderDO, in
      * Update a saml service provider if already exists.
      *
      * @param serviceProviderDO     Service provider information object.
+     * @param currentIssuer         Issuer of the service provider before the update.
      * @param tenantId              Tenant ID.
      * @return True if success.
      * @throws IdentityException Error when updating the SAML service provider.
      */
-    public boolean updateServiceProvider(SAMLSSOServiceProviderDO serviceProviderDO, int tenantId)
+    public boolean updateServiceProvider(SAMLSSOServiceProviderDO serviceProviderDO, String currentIssuer, int tenantId)
```
[the remark this answers is not in the document]
Comment (hwupathum) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/SAMLSSOServiceProviderManager.java:80]: This method has not been used anywhere yet. Therefore changing the method parameters will not have any affect.

**Review by madurangasiriwardena — APPROVED** [review]

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java (hunk 1 above)
Comment (hwupathum) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAO.java:526]: With the current implementation, both delete and add functions happens inside a transaction therefore it won't be an issue with the deletion of resource

**Pull request merged**